Mathlib/RingTheory/PowerSeries/WeierstrassPreparation.lean

```diff
@@ -435,4 +454,192 @@ theorem weierstrassMod_zero_left [IsAdicComplete (IsLocalRing.maximalIdeal A) A]
 
 end IsLocalRing
 
+/-!
+
+## Weierstrass preparation theorem
+
+-/
+
+/-- If `f` is a polynomial over `A`, `g` and `h` are power series over `A`,
+then `PowerSeries.IsWeierstrassFactorizationAt g f h I` is a `Prop` which asserts that `f` is
+distingushed at `I`, `h` is a unit, such that `g = f * h`. -/
+@[mk_iff]
+structure IsWeierstrassFactorizationAt (g : A⟦X⟧) (f : A[X]) (h : A⟦X⟧) (I : Ideal A) : Prop where
+  isDistinguishedAt : f.IsDistinguishedAt I
+  isUnit : IsUnit h
+  eq_mul : g = f * h
+
+/-- Version of `PowerSeries.IsWeierstrassFactorizationAt` for local rings with respect to
+its maximal ideal. -/
+abbrev IsWeierstrassFactorization (g : A⟦X⟧) (f : A[X]) (h : A⟦X⟧) [IsLocalRing A] : Prop :=
+  g.IsWeierstrassFactorizationAt f h (IsLocalRing.maximalIdeal A)
+
+namespace IsWeierstrassFactorizationAt
+
+variable {g : A⟦X⟧} {f : A[X]} {h : A⟦X⟧} {I : Ideal A} (H : g.IsWeierstrassFactorizationAt f h I)
+include H
+
+theorem map_ne_zero_of_ne_top (hI : I ≠ ⊤) : g.map (Ideal.Quotient.mk I) ≠ 0 := by
+  rcases subsingleton_or_nontrivial (A ⧸ I) with h | _
+  · apply (Submodule.Quotient.subsingleton_iff _).1 at h
+    exact False.elim <| hI <| by ext; simp [h]
+  rw [congr(map (Ideal.Quotient.mk I) $(H.eq_mul)), map_mul, ← Polynomial.polynomial_map_coe, ne_eq,
+    (H.isUnit.map _).mul_left_eq_zero]
+  exact_mod_cast f.map_monic_ne_zero (f := Ideal.Quotient.mk I) H.isDistinguishedAt.monic
+
+theorem degree_eq_coe_lift_order_map_of_ne_top (hI : I ≠ ⊤) :
+    f.degree = (g.map (Ideal.Quotient.mk I)).order.lift
+      (order_finite_iff_ne_zero.2 (H.map_ne_zero_of_ne_top hI)) := by
```

Comment:
```suggestion
      (order_finite_iff_ne_zero.mpr (H.map_ne_zero_of_ne_top hI)) := by
```

Reply:
Then all `.2` should be `.mpr`... including several one at `Distinguished.lean`. Personally I like `.1` and `.2` because I can't remember in `.mp` and `.mpr` which is forward and which is backward, just as someone complained in Lean Zulip,,,

Reply:
Don't worry too much, sooner or later we will make a final decision.